### PR TITLE
RefinedDiscrTree

### PR DESCRIPTION
**Pull request by JovanGerb**
This is the same as my DiscrTree PR in Std, but I've decided to move it over to Mathlib now. Additionally, I have added the `rw??` tactic which uses this `RefinedDiscrTree` in combination with shift-clicking on the goal to do targeted library rewriting.

I've made a new version of Lean's `DiscrTree` called `RefinedDiscrTree` with a lot of new features. It is able to index lambda's, forall's, and their bound variables. It also gives library metavariables an index so that it can keep track of which metavariables are the same. It also gives a score to each match pattern, so that you can get the results in order of how relevant they are.

The file uses the `StateListT` monad transformer (which does not satisfy the associativity monad law) that I made, so I also added its definition in another file.

The main purpose is for use is to get better results in library searches, for tactics like exact? and apply?, and I want to publish a library rewrite tactic that uses this discrimination tree.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/LibraryRewrite.lean [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+  return some { rwLemma with tactic, extraGoals, replacement }
+
+
+def renderResults (results : Array (CodeWithInfos × Array RewriteApplication)) (isEverything : Bool)
+    (range : Lsp.Range) (doc : FileWorker.EditableDocument) : Html :=
+  let htmls := results.map (fun (t, arr) => renderBlock t arr)
+  let title := s! "{if isEverything then "All" else "Some"} rewrite suggestions:";
+  <details «open»={true}>
+    <summary className="mv2 pointer"> {.text title}</summary>
+    {.element "div" #[("style", json% {"margin-left" : "4px"})] htmls}
```
Comment (Vtec234) [inline Mathlib/Tactic/LibraryRewrite.lean:245]:
```suggestion
    {.element "div" #[("style", json% {"marginLeft" : "4px"})] htmls}
```
We pass the props to React, so React-style names should be used.

File: Mathlib/Tactic/LibraryRewrite.lean [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+  </details>
+where
+  renderBlock (title : CodeWithInfos) (results : Array RewriteApplication) : Html :=
+    let core := .element "div" #[("style", json% {"margin-left" : "4px"})] <|
```
Comment (Vtec234) [inline Mathlib/Tactic/LibraryRewrite.lean:249]:
```suggestion
    let core := .element "div" #[("style", json% {"marginLeft" : "4px"})] <|
```

File: Mathlib/Tactic/LibraryRewrite.lean [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+      results.map fun rw =>
+        let button := Html.ofComponent MakeEditLink
+              (.ofReplaceRange doc.meta range rw.tactic none)
+              #[.text s! "{rw.name}"]
+        let replacement := <InteractiveCode fmt={rw.replacement}/>
+        let extraGoals := rw.extraGoals.concatMap
+          (#[<br/>, <strong «class»="goal-vdash">⊢ </strong>, <InteractiveCode fmt={·}/>])
```
Comment (Vtec234) [inline Mathlib/Tactic/LibraryRewrite.lean:256]:
```suggestion
          (#[<br/>, <strong className="goal-vdash">⊢ </strong>, <InteractiveCode fmt={·}/>])
```

File: Mathlib/Tactic/LibraryRewrite.lean (hunk 1 above)
Comment (JovanGerb) [reply Mathlib/Tactic/LibraryRewrite.lean]: Ok, I've made the changes

**Comment by JovanGerb** [comment]
This PR is from a fork of Mathlib, so I moved it over here: #11768

**Pull request closed without merging**